src/memray/_memray/macho_shenanigans.cpp

```diff
@@ -0,0 +1,145 @@
+#endif
+
+namespace memray::linker {
+
+template<typename Hook>
+static void
+patch_symbol(
+        const Hook& hook,
+        typename Hook::signature_t intercept,
+        const char* symname,
+        void* addr,
+        vm_size_t section_size,
+        bool restore_original)
+{
+    vm_size_t trunc_address = trunc_page((vm_size_t)addr);
+    vm_size_t trunc_size = (vm_size_t)addr - trunc_address;
+
+    kern_return_t err = vm_protect(
+            mach_task_self(),
+            (uintptr_t)trunc_address,
+            section_size + trunc_size,
+            0,
+            VM_PROT_READ | VM_PROT_WRITE | VM_PROT_COPY);
+    if (err == KERN_SUCCESS) {
+        auto typedAddr = reinterpret_cast<typename Hook::signature_t*>(addr);
+        *typedAddr = restore_original ? hook.d_original : intercept;
+        LOG_IF_DEBUG(symname << " intercepted!");
+    } else {
+        LOG(ERROR) << "Failed to patch " << symname;
+    }
+}
+
+static void
+patch_symbols_in_section(
+        const section_t* section,
+        uintptr_t slide,
+        const DynamicInfoTable& dyninfo_table,
+        bool restore_original)
+{
+    auto symbol_addr_table = reinterpret_cast<void**>(slide + section->addr);
+    for (unsigned int i = 0; i < section->size / sizeof(void*); i++) {
+        const char* symbol_name = dyninfo_table.getSymbol(section->reserved1, i);
+        if (!symbol_name || !symbol_name[0] || !symbol_name[1]) {
+            continue;
+        }
+#define FOR_EACH_HOOKED_FUNCTION(hookname)                                                              \
+    if (strcmp(hooks::hookname.d_symbol, symbol_name + 1) == 0) {                                       \
+        patch_symbol(                                                                                   \
+                hooks::hookname,                                                                        \
+                &intercept::hookname,                                                                   \
+                symbol_name,                                                                            \
+                symbol_addr_table + i,                                                                  \
+                section->size,                                                                          \
+                restore_original);                                                                      \
+        continue;                                                                                       \
+    }
+        MEMRAY_HOOKED_FUNCTIONS
+#undef FOR_EACH_HOOKED_FUNCTION
+    }
+}
+
+static void
+patch_symbols_in_shared_object(const struct mach_header* header, intptr_t slide, bool restore_original)
+{
+    Dl_info info;
+    if (dladdr(header, &info) == 0) {
+        return;
+    }
+
+    const char* image_name = info.dli_fname;
+    if (strstr(image_name, "memray.cpython") || strstr(image_name, "/dyld")
+        || strstr(image_name, "dyld.dylib"))
```

Comment:
How stable are these names? It looks like you're matching "dyld" in two different ways, and I'm surprised to see ".cpython" in the name of the "memray" one - do we have to worry about these changing over time?

Reply:
> and I'm surprised to see ".cpython" in the name of the "memray" one
That's to avoid matching something that just has "memray" on the full path (like a `memray` folder).

> How stable are these names? It looks like you're matching "dyld" 

They are stable across different versions for what I have seen (the CI checks it). This basically avoids patching the linker. One is the linker itself (`/dyld`) and the other is its runtime library (`dyld.dylib`). I didn't want to just add `dyld` to avoid unwanted substring matching

Reply:
Oh, for some reason I was misreading this as though the part after the `.` was some sort of namespace, but of course it's not, and that's just part of our filename. Nevermind this, you're right, this should be stable if it's just based on file names.